staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,895 @@
+	if !k8scache.WaitForCacheSync(stopCh, ca.localAttsInformer.HasSynced) {
+		err = fmt.Errorf("Caches failed to sync")
+	}
+	return
+}
+
+func (ca *ConnectionAgent) processQueue() {
+	for {
+		item, stop := ca.queue.Get()
+		if stop {
+			return
+		}
+		attRef := item.(attQueueRef)
+		ca.processQueueItem(attRef)
+	}
+}
+
+func (ca *ConnectionAgent) processQueueItem(attRef attQueueRef) {
+	defer ca.queue.Done(attRef)
+	var err error
+	var localStr string
+	if attRef.local {
+		localStr = "local"
+		err = ca.processLocalAtt(attRef)
+	} else {
+		localStr = "remote"
+		err = ca.processRemoteAtt(attRef)
+	}
+	requeues := ca.queue.NumRequeues(attRef.nsn)
+	if err != nil {
+		glog.Warningf("Failed processing %s NetworkAttachment %s in VN with ID %d, requeuing (%d earlier requeues): %s\n",
+			localStr,
+			attRef.nsn,
+			attRef.vni,
+			requeues,
+			err.Error())
+		ca.queue.AddRateLimited(attRef)
+		return
+	}
+	glog.V(4).Infof("Finished %s NetworkAttachment %s in VN with ID %d with %d requeues\n", localStr, attRef.nsn, attRef.vni, requeues)
+	ca.queue.Forget(attRef)
+}
+
+func (ca *ConnectionAgent) onLocalAttAdded(obj interface{}) {
+	localAtt := obj.(*netv1a1.NetworkAttachment)
+	glog.V(5).Infof("Notified of NetworkAttachment %#+v addition to local NetworkAttachments cache\n", localAtt)
+	localAttRef := ca.fromAttToEnquableRef(localAtt)
+	ca.queue.Add(localAttRef)
+}
+
+// TODO make the enqueuing conditional on a check on whether the update is relevant (e.g. an update which adds a metadata
+// label is completely irrelevant and should not trigger an enqueuing)
+func (ca *ConnectionAgent) onLocalAttUpdated(oldObj, newObj interface{}) {
+	oldLocalAtt := oldObj.(*netv1a1.NetworkAttachment)
+	newLocalAtt := newObj.(*netv1a1.NetworkAttachment)
+	glog.V(5).Infof("Notified of NetworkAttachment update from %#+v to %#+v in local NetworkAttachments cache\n", oldLocalAtt, newLocalAtt)
+
+	newLocalAttRef := ca.fromAttToEnquableRef(newLocalAtt)
+	ca.queue.Add(newLocalAttRef)
+	if newLocalAtt.Spec.VNI != oldLocalAtt.Spec.VNI {
+		// A reference is made up by a namespaced name, a vni, and a locality flag. Changes to namespace and name are not possible:
+		// that's equivalent to creating a new, independent network attachment. Changes to the node where the network attachment runs
+		// won't result in an update but in a deletion because the informer this handler is registered to has a field selector on the
+		// value of this node. The only changes to a NetworkAttachment that cause an update and a change in the fields of its enquable
+		// reference are those to its VNI value. In this case, the old object and the updated object have different references. We enqueue
+		// both so that not only the new data structures (e.g. Network Interface) are allocated, but the old ones are deallocated (by two
+		// separate worker threads).
+		oldLocalAttRef := ca.fromAttToEnquableRef(oldLocalAtt)
+		ca.queue.Add(oldLocalAttRef)
+	}
```

Comment:
What if the remoteness changed?

Reply:
Please clarify what you mean by "remoteness change".
If you refer to an update which changes a local attachment A1 node, making A1 transition from local to remote,   the case is handled correctly. The informer the handlers whose names begin with "onLocalAtt" are registered to uses a List&Watch option with a field selector like this: "spec.node=<local node>". If an attachment is updated and the update makes the attachment no longer match that selector, the attachment is removed from the Informer cache, and onLocalAttDeleted(obj interface{}) is invoked rather than onLocalAttUpdated(oldObj, newObj interface{}). It's written in the comments above the lines you commented. If you did not find them clear enough I'll change them.

Reply:
Oh, right.  A change in remoteness appears as a delete and a create.  Again, it would be best to sequence the implementation of the delete before the implementation of the create.

Reply:
We agreed no to change the current implementation because we could not come up with a better alternative.